### PR TITLE
[Finishes #99188090] Examples on close

### PR DESCRIPTION
**Pull request by Bjvanminnen**
Run examples on close, and alert if any are wrong.

image

This dialog is pretty similar to the confirm dialog we give you when you click start over, so I made a single showSimpleDialog that is used for both.

I also did some bug fixing with how examples on run were working in Eval.

I also did some refactoring of how examples work in general. The biggest change was to make it so that our test example functions return either an error string or null on success. This allows us to consolidate our success string to be in a single place (in blockly-core).

### CODE REVIEW

File: apps/src/StudioApp.js [hunk 1]
```diff
@@ -1812,6 +1826,30 @@ StudioApp.prototype.getUnfilledFunctionalBlockError = function (topLevelType) {
 };
 
 /**
+ * Looks for failing examples, and updates the result text for them if they're
+ * open in the contract editor
+ * @param {function} failureChecker
```
Comment (bcjordan) [inline apps/src/StudioApp.js]: maybe short note here about what params it takes / what it returns?

File: blockly-core/core/ui/contract_editor/contract_editor.js [hunk 2]
```diff
@@ -394,15 +396,33 @@ Blockly.ContractEditor.prototype.moveExampleBlocksToModal_ = function (functionN
...
+ */
 Blockly.ContractEditor.prototype.registerTestResetHandler = function (testResetHandler) {
   this.testResetHandler_ = testResetHandler;
 };
 
 /**
+ * Allows app to pass in a function that is called on contract editor close if
+ * any examples fail. This function should return true if the app wants to own
+ * closing the dialog
+ */
```
Comment (bcjordan) [inline blockly-core/core/ui/contract_editor/contract_editor.js]: could use `@param {function}` annotation here, maybe mentioning return true/false meaning

File: blockly-core/core/ui/contract_editor/contract_editor.js [hunk 3]
```diff
@@ -394,15 +396,33 @@ Blockly.ContractEditor.prototype.moveExampleBlocksToModal_ = function (functionN
...
+Blockly.ContractEditor.prototype.registerTestsFailedOnCloseHandler = function (handler) {
+  this.appHijackedDialogClose_ = handler;
```
Comment (bcjordan) [inline blockly-core/core/ui/contract_editor/contract_editor.js]: Wondering what might be a good name for this to make its behavior a little more clear.

The callback kind of fills two roles, (1) return true/false if it should subsume the closing responsibility, (2) execute custom logic on a failure's close.

One option would be to, s/`this.appsHijackedDialogClose_/`this.customFailureCloseHandler_`, and if `this.customFailureCloseHandler_` is null, close normally on fail. Then the existence handles responsibility #1, and the callback has one less job.

File: blockly-core/core/ui/contract_editor/contract_editor.js [hunk 4]
```diff
@@ -815,10 +835,12 @@ Blockly.ContractEditor.prototype.resetExampleViews = function () {
 
 /**
  * Call our app-specific test handler for this block
- * @return {string} Result describing pass/failure
+ * @param {Blockly.Block}
+ * @param {boolean} True if app should visualize the test
```
Comment (bcjordan) [inline blockly-core/core/ui/contract_editor/contract_editor.js]: s/`{boolean}`/`{boolean} visualize`

File: blockly-core/core/ui/contract_editor/contract_editor.js [hunk 5]
```diff
@@ -815,10 +835,12 @@ Blockly.ContractEditor.prototype.resetExampleViews = function () {
 
 /**
  * Call our app-specific test handler for this block
- * @return {string} Result describing pass/failure
+ * @param {Blockly.Block}
```
Comment (bcjordan) [inline blockly-core/core/ui/contract_editor/contract_editor.js]: `{Blockly.Block} block`

File: blockly-core/core/ui/contract_editor/contract_editor.js [hunk 6]
```diff
@@ -921,3 +943,23 @@ Blockly.ContractEditor.prototype.onPlaceExampleContent = function (currentY) {
...
+    view.refreshTestingUI(false);
+    allPass = allPass && !failure;
+  }.bind(this));
+  if (!allPass && this.appHijackedDialogClose_()) {
+    // app has taken responsibilty for closing dialog (likely by launching
+    // modal confirm dialog
+    return;
+  }
+
+  this.hideIfOpen.call(this);
```
Comment (bcjordan) [inline blockly-core/core/ui/contract_editor/contract_editor.js]: Is there a reason this isn't `this.hideIfOpen()`?

**Comment by bcjordan** [comment]
LGTM overall, just a few nits. I like the flipping of example evaluation result string meaning.

File: blockly-core/core/ui/contract_editor/contract_editor.js (hunk 6 above)
Comment (Bjvanminnen) [inline blockly-core/core/ui/contract_editor/contract_editor.js]: Heh. Probably not. I think my first approach had me sending params to hideIfOpen.

**Pull request merged**